### DISCUSSION TITLE
Error when seeding

### DISCUSSION THREAD

**Discussion by Tobbe** [prisma/prisma#7123] [Q&A] [answered]
I'm using RedwoodJS and I'm trying to seed my DB. But I get this error
```
$ yarn rw prisma db seed
yarn run v1.22.4
$ C:\Users\tobbe\dev\redwood\performance-management\node_modules\.bin\rw prisma db seed

Running Prisma CLI:
yarn prisma db seed --preview-feature --schema "C:\Users\tobbe\dev\redwood\performance-management\api\prisma\schema.prisma"

Prisma schema loaded from prisma\schema.prisma
Running node "prisma\seed.js" ...
PrismaClientKnownRequestError2 [PrismaClientKnownRequestError]: The change you are trying to make would violate the required relation 'PlayerToPosition' between the `Player` and `Position` models.
    at cb (C:\Users\tobbe\dev\redwood\performance-management\node_modules\@prisma\client\runtime\index.js:34862:17)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
    at async main (C:\Users\tobbe\dev\redwood\performance-management\api\prisma\seed.js:20:3) {
  code: 'P2014',
  clientVersion: '2.20.1',
  meta: {
    relation_name: 'PlayerToPosition',
    model_a_name: 'Player',
    model_b_name: 'Position'
  }
}

Your database has been seeded.
Done in 7.80s.
```
These are the two relevant models
```prisma
model Player {
  id            Int             @id @default(autoincrement())
  position      Position        @relation(fields: [positionId], references: [id])
  positionId    String
  PlayerToSkill PlayerToSkill[]
  UserToPlayer  UserToPlayer[]
}

model Position {
  id               String             @id @default(uuid())
  number           Int
  nameNh           String
  nameSh           String
  players          Player[]
  SkillToPosition  SkillToPosition[]
  defaultSkills    DefaultSkill[]
}
```
I don't understand why the error message is mentioning a `PlayerToPosition` table. Can someone please explain? 

Thanks!

**Comment by Tobbe** [comment]
Here's (a simplified version of) my seed.js file
```js
/* eslint-disable no-console */
const { PrismaClient } = require('@prisma/client')
const dotenv = require('dotenv')

dotenv.config()
const db = new PrismaClient()

async function main() {
  // TRUNCATE TABLE "Position"
  await db.position.deleteMany({ where: {} })

  await db.position.createMany({
    data: [
      {
        id: 'f858e875-076f-4da4-b933-2845d39f6b96',
        number: 1,
        nameNh: 'Loose Head Prop',
        nameSh: 'Loose Head Prop',
      },
    ],
  })
}

main()
  .catch((e) => console.error(e))
  .finally(async () => {
    await db.$disconnect()
  })
```

**Reply by Tobbe** [reply]
Removing the TRUNCATE from my seed script makes it go through. But I'd still like to understand this better. Why did it think a join table was needed between Player and Position?

**Reply by ryands17** [reply]
@Tobbe 👋 

To understand correctly, does it work properly when you remove the truncate and replace it with `deleteMany`?

**Reply by Tobbe** [reply]
Ohh, sorry, the `deleteMany` is the `TRUNCATE`. I never did try it with raw sql, I just put that comment there for future me to know what the weird looking `deleteMeny({ where: {}})` was all about

So I have to do this for it to "work"
```
  // TRUNCATE TABLE "Position"
  // await db.position.deleteMany({ where: {} })
```

What I've done instead for now is to use an upsert to make sure I overwrite any changes to existing positions, and create those that are missing. This still leaves any extra rows in the table though that aren't being used and aren't needed. And it's a lot of duplicated code (same data in both the update and the insert).

To be clear though, Prisma is right in saying "The change you are trying to make would violate the required relation" when I try to truncate the table. So I understand why I'm not allowed to do that. What threw me off though is the mention of `PlayerToPosition`, which is a join table I only thought would be implicitly created for m-n relations.

**Reply by ryands17** [reply]
Most likely you are facing this issue: https://github.com/prisma/prisma/issues/2328

We are actively working on this so it would be great if you could follow the above issue for more updates :)

**Reply by Tobbe** [reply]
Thanks for helping me @ryands17. I've subscribed to the issue you mentioned 🔖 

But the more I think about this, the more I think it's just a confusing error message from Prisma.

So I saw this message "The change you are trying to make would violate the required relation 'PlayerToPosition' between the `Player` and `Position` models." and I was like "What did I do wrong to make Prisma think it should create a m-n join table for me?" And I know Prisma uses the "A" and "B" column names in the tables it creates for implicit m-n relations, so this:
```
    model_a_name: 'Player',
    model_b_name: 'Position'
```
further made me believe Prisma was thinking I wanted an m-n relation between players and positions.

But, now that I've looked at this for a while, I think Prisma is actually building an internal model where it names the 1-n relation "PlayerToPosition", and that's what my `deleteMany` would violate. Is that correct?

**Reply by ryands17** [reply, marked as answer]
Ahh I misunderstood. This is 1-many relation and not a many-to-many. In this case, the `deleteMany` should work if the relation is optional.

So in this schema:

```prisma
model Player {
  id         Int       @id @default(autoincrement())
  position   Position? @relation(fields: [positionId], references: [id])
  positionId String?
}

model Position {
  id      String   @id @default(uuid())
  number  Int
  nameNh  String
  nameSh  String
  players Player[]
}
```

The issue still lies with Cascade deletes and you can track that here as well: https://github.com/prisma/prisma/issues/2057

**Reply by Tobbe** [reply]
Thanks. But is my "guess" above correct? (Quoted again below, for reference)
> I think Prisma is actually building an internal model where it names the 1-n relation "PlayerToPosition", and that's what my deleteMany would violate. Is that correct?

**Reply by ryands17** [reply]
This is just the default relation name that you can customise via the `@relation` attribute as [follows](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#disambiguating-relations). If you explicitly specify the relation name, that will be used as that relation is currently being violated.
